File: moderate/prime_less.r
primes <- c(2, 3, 5, 7, 11, 13)
isPrime <- function(a) {
  i <- 1
  while (a %% primes[i] > 0) {
    if (primes[i] * primes[i] > a) {
      primes <<- append(primes, a)
      return(TRUE)
    }
    i <- i + 1
  }
  FALSE
}

primeSeq <- function() {
  p0 <- 1
  i <- 0
  return(function() {
    if (p0 <= length(primes)) {
      i <<- primes[p0]
      p0 <<- p0 + 1
      return(i)
    }
    i <<- i + 2
    while (!isPrime(i)) {
      i <<- i + 2
    }
    p0 <<- p0 + 1
    return(i)
  })
}

primeless <- function(n) {
  r <- c()
  nextPrime <- primeSeq()
  i <- nextPrime()
  while (i < n) {
    r <- append(r, i)
    i <- nextPrime()
  }
  paste(r, collapse=",")
}

for (line in readLines(tail(commandArgs(), n=1))) {
  cat(sapply(as.integer(line), primeless), sep="\n")
}

File: moderate/prime_less.lua
primes = {2, 3, 5, 7, 11, 13}
function isPrime(a)
  local b = 1
  while a % primes[b] > 0 do
    if primes[b] * primes[b] > a then
      primes[#primes + 1] = a
      return true
    end
    b = b + 1
  end
  return false
end

function primeSeq()
  local p0, i = 1, 0
  return function()
           if p0 <= #primes then
             i, p0 = primes[p0], p0+1
             return i
           end
           i = i + 2
           while not isPrime(i) do
             i = i + 2
           end
           p0 = p0 + 1
           return i
         end
end

for line in io.lines(arg[1]) do
  local n, nextPrime = tonumber(line), primeSeq()
  local i = nextPrime()
  while i < n do
    if i > 2 then io.write(",") end
    io.write(i)
    i = nextPrime()
  end
  print()
end
